src/sp/it/pl/core/CoreMenus.kt

```diff
@@ -51,177 +49,134 @@ class CoreMenus: Core {
-                                }
-                            }
+                    item("Copy to clipboard") { copyToSysClipboard(DataFormat.IMAGE, selected.image) }
+                }
+            if (!selected.fsDisabled)
+                menu("Image file") {
+                    item("Browse location") { selected.fsImageFile.browse() }
+                    item("Open (in associated program)") { selected.fsImageFile.open() }
+                    item("Edit (in associated editor)") { selected.fsImageFile.edit() }
+                    item("Delete from disc") { selected.fsImageFile.recycle() }
+                    item("Fullscreen") {
+                        val f = selected.fsImageFile
+                        if (ImageFileFormat.isSupported(f)) {
+                            APP.actions.openImageFullscreen(f)
                         }
-                    },
-                    runIf(!cmd.fsDisabled) {
-                        menu("Image file") {
-                            item("Browse location") { cmd.fsImageFile.browse() }
-                            item("Open (in associated program)") { cmd.fsImageFile.open() }
-                            item("Edit (in associated editor)") { cmd.fsImageFile.edit() }
-                            item("Delete from disc") { cmd.fsImageFile.recycle() }
-                            item("Fullscreen") {
-                                val f = cmd.fsImageFile
-                                if (ImageFileFormat.isSupported(f)) {
-                                    APP.actions.openImageFullscreen(f)
-                                }
-                            }
-                        }
-                    },
-                    runIf(cmd.representant!=null) {
-                        menuOfItemsFor(contextMenu, cmd.representant)
                     }
-            )
+                }
+            if (selected.representant!=null)
+                addMenuOfItemsFor(contextMenu, selected.representant)
         }
 
     }
 
     companion object {
 
-        /** DSL for creating menu. */
-        private fun <T: MenuItem> menuItems(vararg elements: T?) = seqOf(*elements).filterNotNull().filterNot { it is Menu && it.items.isEmpty() }
+        private inline fun <reified W> Menu.widgetItems(crossinline action: (W) -> Unit) =
+                items(APP.widgetManager.factories.getFactoriesWith<W>(),
+                        { it.nameGui() },
+                        { it.use(WidgetSource.NO_LAYOUT) { action(it) } })
 
-        /** DSL for creating menu. */
-        private inline fun <reified W> Menu.widgetItems(crossinline action: (W) -> Unit) = items(
-                APP.widgetManager.factories.getFactoriesWith<W>(),
-                { it.nameGui() },
-                { it.use(WidgetSource.NO_LAYOUT) { action(it) } }
-        )
+        private inline fun <reified W> ContextMenuBuilder<*>.widgetMenu(text: String, crossinline action: (W) -> Unit) =
```

Comment:
Convenience method. It does spare a line. But!

I purposefully did not define this method as to avoid one liner for widget context menus.

```
 widgetMenu<SongReader>("Show in") { it.read(selected.items) }
```
```
menu("Show in") {
    widgetItems<SongReader> { it.read(pig.items) }
}
```
Paradoxically, the latter is to be preferred!

Why?
- it is more declarative (mirrors the structure of the resulting contextMenu better). Skimming the code already reveals that there is nesting involved. The one liner does no such thing!
- it hides the menu instance, i.e. it is not possible to add more items to it, which is actually what we need to do in some cases ("Explore in"). Using the separate method in such cases would lead to more complex API

A symptom that something is wrong here is how the widgetMenu() method 'sticks out'. Compare:
```
            item("Play items") { PlaylistManager.use { it.playItem(selected.items[0]) } }
            item("Remove items") { PlaylistManager.use { it.removeAll(selected.items) } }
            widgetMenu<SongReader>("Show in") { it.read(selected.items) }
            widgetMenu<SongWriter>("Edit tags in") { it.read(selected.items) }
            item("Crop items") { PlaylistManager.use { it.retainAll(selected.items) } }
            item("Duplicate items as group") { PlaylistManager.use { it.duplicateItemsAsGroup(selected.items) } }
```
```
            item("Play items") { PlaylistManager.use { it.playItem(selected.items[0]) } }
            item("Remove items") { PlaylistManager.use { it.removeAll(selected.items) } }
            menu("Show in") { .............................. it.read(selected.items) }
            menu("Edit tags in") { .......................... it.read(selected.items) }
            item("Crop items") { PlaylistManager.use { it.retainAll(selected.items) } }
            item("Duplicate items as group") { PlaylistManager.use { it.duplicateItemsAsGroup(selected.items) } }
```
It is a case of over-engineering, because the menu itself is fine being constructed as `menu("name")`. It is the menu's items that are 'special' and require convenience method. You took it beyond its responsible abstraction layer. Why pollute menus with how items are constructed. It is shorter, yes, but it is not correct, we spare lines but lose flexibility. 

There is a point at which convenience method does more harm than use and it is when it limits the flexibility of the API.

Reply:
okay